src/lib.rs

```diff
@@ -293,43 +293,64 @@ pub fn count(i: usize) -> usize {
 
 /// Returns all the previous fully rooted trees before the node.
 ///
+/// ## Panics
+/// If an uneven index is passed.
+///
 /// ## Examples
 /// ```rust
-/// assert_eq!(flat_tree::full_roots(0), []);
-/// assert_eq!(flat_tree::full_roots(2), [0]);
-/// assert_eq!(flat_tree::full_roots(8), [3]);
-/// assert_eq!(flat_tree::full_roots(20), [7, 17]);
-/// assert_eq!(flat_tree::full_roots(18), [7, 16]);
-/// assert_eq!(flat_tree::full_roots(16), [7]);
+/// use flat_tree::full_roots;
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(0, &mut nodes);
+/// assert_eq!(nodes, []);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(2, &mut nodes);
+/// assert_eq!(nodes, [0]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(8, &mut nodes);
+/// assert_eq!(nodes, [3]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(20, &mut nodes);
+/// assert_eq!(nodes, [7, 17]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(18, &mut nodes);
+/// assert_eq!(nodes, [7, 16]);
+///
+/// let mut nodes = Vec::with_capacity(16);
+/// full_roots(16, &mut nodes);
+/// assert_eq!(nodes, [7]);
 /// ```
-pub fn full_roots(i: usize) -> Vec<usize> {
-  let mut result = Vec::with_capacity(64);
-
-  if is_odd(i) {
-    result
-  } else {
-    let mut tmp = i >> 1;
-    let mut offset = 0;
-    let mut factor = 1;
-
-    loop {
-      if tmp == 0 {
-        break;
-      }
-      while factor * 2 <= tmp {
-        factor *= 2;
-      }
-      result.push(offset + factor - 1);
-      offset += 2 * factor;
-      tmp -= factor;
-      factor = 1;
+pub fn full_roots(i: usize, nodes: &mut Vec<usize>) {
+  assert!(
+    is_even(i),
+    format!(
+      "You can only look up roots for depth 0 blocks, got index {}",
+      i
+    )
+  );
+  let mut tmp = i >> 1;
+  let mut offset = 0;
+  let mut factor = 1;
+
+  loop {
+    if tmp == 0 {
+      break;
     }
-
-    result
+    while factor * 2 <= tmp {
+      factor *= 2;
+    }
+    nodes.push(offset + factor - 1);
+    offset += 2 * factor;
+    tmp -= factor;
+    factor = 1;
   }
 }
 
-#[inline(always)]
+#[inline]
 fn is_even(num: usize) -> bool {
```

Comment:
What's the difference between them?

Reply:
`clippy` doesn't like `[inline(always)]` because it hard overrides compiler behavior. Best to think of `[inline]` as a hint to the compiler: tell it that it should probably inline this function, but leave the final decision up to it.

Think it's probably reasonable to delegate this type of decision making as much as possible to LLVM, but put hints in there where it makes sense. Think that's also why clippy flags the forceful approach.

Reply:
Gotcha. The compiler probably knows best most of the time :smile: